### PR TITLE
feat: pipe for Sonatype Nexus Raw repositories

### PR DESCRIPTION
**Pull request by antonienko**
Enables Goreleaser to upload artifacts to [Sonatype Nexus](https://www.sonatype.com/product-nexus-repository) Raw repositories.

### CODE REVIEW

**Comment by antonienko** [comment]
I needed this feature for my own selfish goals, and I figured I could contribute it to the project. Since it's my first contribution, I'm open to receive feedback and suggestions on how to improve this pull request.

File: internal/pipe/nexus_raw/nexus_raw.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+		body, err := ioutil.ReadAll(res.Body)
+		if err == nil {
+			bodyContents = string(body)
+		}
+		return pipe.Skip(fmt.Sprintf("bad status: %s, body: %s", res.Status, bodyContents))
+	}
+	return nil
+}
+
+func createForm(w *multipart.Writer, artifPath, artifName, nexusDir string) error {
```
Comment (caarlos0) [inline internal/pipe/nexus_raw/nexus_raw.go]: errors inside this function shouldn't be `skip errors`.

skip errs are used only to skip a pipe if some configuration is missing, other that that, regular errors should be used.

File: pkg/config/config.go [hunk 2]
```diff
@@ -340,6 +340,14 @@ type Put struct {
 	Signature      bool     `yaml:",omitempty"`
 }
 
+type NexusRaw struct {
+	Url        string `yaml:",omitempty"`
```
Comment (caarlos0) [inline pkg/config/config.go]:
```suggestion
	URL        string `yaml:",omitempty"`
```

File: pkg/config/config.go [hunk 3]
```diff
@@ -340,6 +340,14 @@ type Put struct {
 	Signature      bool     `yaml:",omitempty"`
 }
 
+type NexusRaw struct {
```
Comment (caarlos0) [inline pkg/config/config.go]: missing struct docs here...

File: internal/pipe/nexus_raw/nexus_raw.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
+package nexus_raw
```
Comment (caarlos0) [inline internal/pipe/nexus_raw/nexus_raw.go]: missing package docs

File: internal/pipe/nexus_raw/nexus_raw.go [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+func (Pipe) Publish(ctx *context.Context) error {
+	if len(ctx.Config.NexusRaws) == 0 {
+		return pipe.Skip("sonatype nexus section is not configured")
+	}
+	for _, instance := range ctx.Config.NexusRaws {
+		if skipErr := checkConfig(&instance); skipErr != nil {
+			return pipe.Skip(skipErr.Error())
+		}
```
Comment (caarlos0) [inline internal/pipe/nexus_raw/nexus_raw.go]: could be just:

```go
for _, instance := range ctx.Config.NexusRaws {
		instance := instance
		if err := checkConfig(&instance); err != nil {
			return err
		}
}
```

**Comment by caarlos0** [comment]
Just a couple of comments. :) 

Also would be nice if you could add the docs as well 🚀 

Thanks for the PR 👏

**Comment by antonienko** [comment]
I have added some basic documentation based on the artifactory one.

Since the functionality of this pipe is so simple, I haven't go into details of the different configuration parameters. Let me know if you need anything else at this stage, or we can evolve the documents at the same time that this pipe grows.

**Comment by antonienko** [comment]
@caarlos0 Is there anything else I need to do for the pull request to be reviewed?

**Comment by caarlos0** [comment]
> @caarlos0 Is there anything else I need to do for the pull request to be reviewed?

no, I just haven't make the time to look into it yet... sorry for the delay :(

File: internal/pipe/nexusraw/nexusraw.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
+// Package nexus_raw provides a Pipe that push artifacts to a Sonatype Nexus repository of type 'raw'
```
Comment (caarlos0) [inline internal/pipe/nexusraw/nexusraw.go]: `nexus_raw` -> `nexusraw`

File: internal/pipe/nexusraw/nexusraw.go [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+		return fmt.Errorf("cannot fill directory field contents: '%s'", err.Error())
+	}
+	filenameField, err := w.CreateFormField("raw.asset1.filename")
+	if err != nil {
+		return fmt.Errorf("cannot create filename field: '%s'", err.Error())
+	}
+	if _, err = io.Copy(filenameField, strings.NewReader(artifName)); err != nil {
+		return fmt.Errorf("cannot fill filename field contents: '%s'", err.Error())
+	}
+	return w.Close()
```
Comment (caarlos0) [inline internal/pipe/nexusraw/nexusraw.go]: shouldn't this be on a `defer`?

File: internal/pipe/nexusraw/nexusraw_test.go [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+	assert.ElementsMatch(t, cs.Items, []item{
+		{"testDirectory/artifact.tar.gz"},
+		{"testDirectory/artifact2.tar.gz"}},
+	)
+}
+
+func randomListen(t *testing.T) string {
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err)
+	listener.Close()
```
Comment (caarlos0) [inline internal/pipe/nexusraw/nexusraw_test.go]: shouldn't it be in a defer?

File: www/content/nexusraw.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+and your credentials (`username` and `password`):
+
+```yaml
+nexus_raws:
+  -
+    url: https://<Your-Instance>:<Your-Port>
+    repository: example-repo
+    directory: example-folder
+    username: your-username
+    password: your-password
```
Comment (caarlos0) [inline www/content/nexusraw.md]: we should probably get the password from an environment variable?

Comment (antonienko) [reply www/content/nexusraw.md]: In order to do this, shoud I use the `EnvFiles` struct in the `config/config.go` file together with the loadEnv function?

Seems like it is very tailor-made to the git release process.

Should I load the env variable in my pipe instead?

Comment (caarlos0) [reply www/content/nexusraw.md]: take a look on the artifactory pipe, it uses the internal `http` package, which maybe you can try to use here as well, which should handle secrets and etc

https://goreleaser.com/artifactory/

Comment (antonienko) [reply www/content/nexusraw.md]: The problem I see with it is that the internal http package has a hardcoded PUT method, so in order to use it I would need to refactor it to be able to use POST, which is the verb that Nexus accepts for uploads.

Comment (caarlos0) [reply www/content/nexusraw.md]: I think we can refactor that first then? Maybe in another PR even?

Basically adding a `method` option which defaults to `PUT`, right? Or are there more things?

Comment (antonienko) [reply www/content/nexusraw.md]: I've been checking out the http package, to assess the refactor to include POST, and I see that PUT is highly hardcoded in it, it wouldn't be a trivial refactor.

I'm sorry to say I don't have time enough to attack this challenge at the moment. 

Would you mind if I finish this Nexus feature loading the environment variable from inside my pipe, and I can hopefully tackle the refactoring of the http package later on?

Comment (caarlos0) [reply www/content/nexusraw.md]: Sorry, I rather not.

I can try to take a look at the `http` package as well...

Comment (caarlos0) [reply www/content/nexusraw.md]: @antonienko would https://github.com/goreleaser/goreleaser/pull/1246 work?

Comment (antonienko) [reply www/content/nexusraw.md]: On a first quick review, it looks great. As soon as it is merged I'll try and complete this PR using it.

Comment (caarlos0) [reply www/content/nexusraw.md]: merged to master :)

**Pull request closed without merging**